tempesta_fw/http.c

```diff
@@ -290,11 +294,42 @@ tfw_http_conn_msg_alloc(TfwConnection *conn)
 		return NULL;
 
 	hm->conn = conn;
+	tfw_connection_get(conn);
 	tfw_gfsm_state_init(&hm->msg.state, conn, TFW_HTTP_FSM_INIT);
 
 	return (TfwMsg *)hm;
 }
 
+/*
+ * Free an HTTP message.
+ * Also, free the connection structure if there's no more references.
+ *
+ * This function should be used anytime when there's a chance that
+ * a connection structure may belong to multiple messages, which is
+ * almost always. If a connection is suddenly closed then it still
+ * can be safely dereferenced and used in the code.
+ * In rare cases we're sure that a connection structure in a message
+ * doesn't have multiple users. For instance, when an error response
+ * is prepared and sent by Tempesta, that HTTP message does not need
+ * a connection structure. The message is then immediately destroyed,
+ * and a simpler tfw_http_msg_free() can be used for that.
+ */
+void
+tfw_http_conn_msg_free(TfwHttpMsg *hm)
+{
+	TfwConnection *msg_conn = NULL;
+
+	if (hm) {
+		msg_conn = hm->conn;
+		tfw_http_msg_free(hm);
+	}
+	if (msg_conn && tfw_connection_put(msg_conn)) {
+		TFW_CONN_TYPE(msg_conn) & Conn_Clnt
+			? tfw_cli_conn_free(msg_conn)
+			: tfw_srv_conn_free(msg_conn);
+	}
```

Comment:
The function could be bit simplier:

        if (unlikely(!hm))
                return;
        
        if (tfw_connection_put(hm->conn)) {
                TFW_CONN_TYPE(msg_conn) & Conn_Clnt
                        ? tfw_cli_conn_free(hm->conn)
                        : tfw_srv_conn_free(hm->conn);
        }
        
        tfw_http_msg_free(hm);